Draft: Implements Encoder-Decoder Attention Model

mmz33 opened the pull request.

mmz33 committed:
implemented enc-dec-att model

mmz33 committed:
fix docs

mmz33 committed:
fix docs

JackTemaki commented:
I would recommend to build full setups before merging this to avoid the same problems we had with other code that we merged but never "used" before, I would help with this.

mmz33 commented:
> I would recommend to build full setups before merging this to avoid the same problems we had with other code that we merged but never "used" before, I would help with this.

Yeah I agree.

mmz33 committed:
better check

mmz33 committed:
better comment

mmz33 committed:
better comment

mmz33 committed:
refactor + implement zoneout

mmz33 committed:
implement zoneout lstm cell

mmz33 committed:
fix

mmz33 committed:
put tensors on cuda

mmz33 committed:
make device configurable

mmz33 committed:
check for cuda availability

JackTemaki committed:
remove explicit device, fix return value

JackTemaki committed:
remove device from attention test

JackTemaki committed:
remove leftover print statement

JackTemaki committed:
Add shift_embeddings flag for Attention Decoder

Allows to pass the label unshifted for step-wise search
without needing a separate function besides "forward".

mmz33 committed:
fix logits dropout bug

JackTemaki commented:
I completely forgot to test with this branch again after using it already for some time, did so now and it works normally.

Atticus1806 commented on i6_models/decoder/attention.py:
```diff
@@ -0,0 +1,209 @@
+        labels: torch.Tensor,
+        enc_seq_len: torch.Tensor,
+        state: Optional[Tuple[torch.Tensor, ...]] = None,
+        shift_embeddings=True,
```
```suggestion
        shift_embeddings: bool = True,
```

JackTemaki committed:
Update i6_models/decoder/attention.py

Co-authored-by: Benedikt Hilmes <hilmes@hltpr.rwth-aachen.de>

JackTemaki commented on i6_models/decoder/zoneout_lstm.py:
```diff
@@ -0,0 +1,47 @@
+from typing import Tuple
+
+
+class ZoneoutLSTMCell(nn.Module):
```
I do not like non-standard types in a config object, and nn.RNNCellBase would be one, thus no, I at least do not want one.

JackTemaki committed:
Disable autocast for ZoneoutLSTM cell

albertz commented on i6_models/decoder/zoneout_lstm.py:
```diff
@@ -0,0 +1,48 @@
+    def forward(
+        self, inputs: torch.Tensor, state: Tuple[torch.Tensor, torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        with torch.autocast(device_type="cuda", enabled=False):
```
Why is this disabled here? That should be explained in the code, maybe with some ref.

SimBe195 committed:
Update attention decoder with my custom implementation and add more tests

SimBe195 committed:
Merge branch 'main' into zeineldeen_att_decoder

SimBe195 committed:
Formatting
